Add Term.capture to capture consumed arguments

diml opened the pull request:
This PR implements the idea described in #101.

I added a test for the new feature, it captures the arguments `-a/--aaa`, `-b/--bbb` and positional arguments:

```
$ ./_build/test/test_capture.native -a 
-a
$ ./_build/test/test_capture.native --aaa
--aaa
$ ./_build/test/test_capture.native -b 0 -a
-a -b 0
$ ./_build/test/test_capture.native -b 0 x y z -a
-a -b 0 x y z
$ ./_build/test/test_capture.native -b 0 --other x y z -a
-a -b 0 x y z
```

dbuenzli commented on src/cmdliner.mli:
```diff
@@ -244,6 +244,11 @@ module Term : sig
   (** [choice_names] is a term that evaluates to the names of the terms
       to choose from. *)
 
+  val capture : 'a t -> ('a * string list) t
```
I would like to find a better name for this. `Term.capture t` is a bit odd to read: we are not capturing the term but the arguments it concretely parsed. 

Here are a few ideas `Term.{captured,parsed,actual,concrete}[_args}`. This might still be confused with accessing the term's actual specification though. Maybe we should throw an `eval` in there. Something like `Term.captured_eval_args`, bof, or maybe `Term.with_evaluated_args` ?

diml replied:
`with_evaluated_args` seems good

diml replied:
I wonder if it doesn't clash with the evaluation of a term though, the documentation looks a bit odd: `... is a term whose evaluation returns the arguments evaluated ...`. Other suggestions: `with_user_specified_args`, `with_args_from_command_line`.

dbuenzli commented:
Also @diml due to its unique position I maintain multiple build systems for `cmdliner`. I added a `jbuild` file a long time ago (apparently that never made it in a release), would you be kind enough to upgrade the thing to whatever `dune` consumes nowadays ? (Feel free to add it to this PR).

diml commented on src/cmdliner_cline.mli:
```diff
@@ -17,6 +17,9 @@ val create :
 val pos_arg : t -> Cmdliner_info.arg -> string list
 
+(** Actual command line arguments from the command line *)
+val actual_args : t -> Cmdliner_info.arg -> string list
```
Sure, done

diml commented:
Sure, no problem. I also renamed the `opam` file to `cmdliner.opam` as currently dune uses `<pkg>.opam` files to detect what packages are declared. I also updated the opam file to opam 2.0. Should I update the opam file to use the dune build by default? It speeds up the build: 2s vs 10s on my machine.

diml commented:
@dbuenzli did you have chance to look at this? I'm happy to split out the build system changes in another PR if you prefer.

dbuenzli commented:
Ah no forgot about it sorry. Will try to merge this in the upcoming days.

dbuenzli commented:
Thanks your patches are in as 96eaaa034ce1 and ddd71ae9b03f. In the end I prefered `Term.with_used_args` over `Term.with_evaluated_args` see 6866e3137fb74098e. 

> Should I update the opam file to use the dune build by default?

In fact I think I will rather update it to use the bootstrap build system to make sure it's properly working well for system packagers.

dbuenzli commented:
> In fact I think I will rather update it to use the bootstrap build system to make sure it's properly working well for system packagers.

A good idea, you actually introduced a regression in 96eaaa034ce128a1ae373c104150cd31f4fe4097, see eb29b3f17b67a85d.

diml commented:
Ok, BTW I believe dune is now widely packaged and has no dependencies apart from OCaml itself, so it's generally an acceptable dependency even for system packages.

dbuenzli commented:
> Ok, BTW I believe dune is now widely packaged and has no dependencies apart from OCaml itself, so it's generally an acceptable dependency even for system packages.

Last time I interacted with people doing system packages they still had strong policies against vendoring, so dune would likely depend on cmdliner (e.g. on debian).
